pubnub-hyper/src/transport/hyper/pubsub.rs

```diff
@@ -85,28 +85,135 @@ impl TransportService<request::Subscribe> for Hyper {
+    let i = if i.is_null() { 0 } else { i.as_u32()? };
+    Some(match i {
+        0 => message::Type::Publish,
+        1 => message::Type::Signal,
+        2 => message::Type::Objects,
+        3 => message::Type::Action,
+        i => message::Type::Unknown(i),
+    })
+}
+
+fn parse_message_route(route: &json::JsonValue) -> Result<Option<message::Route>, ()> {
+    if route.is_null() {
+        return Ok(None);
+    }
+    let route = route.as_str().ok_or(())?;
+    // First try parsing as wildcard, it is more restrictive.
+    if let Ok(val) = route.parse() {
+        return Ok(Some(message::Route::ChannelWildcard(val)));
+    }
+    // Then try parsing as regular name for a channel group.
+    if let Ok(val) = route.parse() {
+        return Ok(Some(message::Route::ChannelGroup(val)));
+    }
+    Err(())
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum ParseMessageError {
+    Type,
+    Route,
+    Channel,
+    Timetoken,
+    SubscribeKey,
+}
+
+fn parse_message(message: &json::object::Object) -> Result<Message, ParseMessageError> {
     let message = Message {
-        message_type: Type::from_json(&message["e"]),
-        route: message["b"].as_str().map(|s| s.parse().ok())?,
-        channel: message["c"].as_str()?.parse().ok()?,
+        message_type: parse_message_type(&message["e"]).ok_or(ParseMessageError::Type)?,
+        route: parse_message_route(&message["b"]).map_err(|_| ParseMessageError::Route)?,
+        channel: message["c"]
+            .as_str()
+            .ok_or(ParseMessageError::Channel)?
+            .parse()
+            .map_err(|_| ParseMessageError::Channel)?,
         json: message["d"].clone(),
         metadata: message["u"].clone(),
         timetoken: Timetoken {
-            t: message["p"]["t"].as_str()?.parse().ok()?,
+            t: message["p"]["t"]
+                .as_str()
+                .ok_or(ParseMessageError::Timetoken)?
+                .parse()
+                .map_err(|_| ParseMessageError::Timetoken)?,
             r: message["p"]["r"].as_u32().unwrap_or(0),
         },
-        client: message["i"].as_str().map(str::to_string),
-        subscribe_key: message["k"].to_string(),
+        client: message["i"].as_str().map(std::borrow::ToOwned::to_owned),
+        subscribe_key: message["k"]
+            .as_str()
+            .ok_or(ParseMessageError::SubscribeKey)?
+            .to_owned(),
         flags: message["f"].as_u32().unwrap_or(0),
     };
-    Some(message)
+    Ok(message)
+}
+
+fn process_subscribe_to(to: &[pubsub::SubscribeTo]) -> (String, String) {
+    let channels = to.iter().filter_map(|to| {
+        to.as_channel()
+            .map(AsRef::<str>::as_ref)
+            .or_else(|| to.as_channel_wildcard().map(AsRef::<str>::as_ref))
+    });
+    let channel_groups = to.iter().filter_map(pubsub::SubscribeTo::as_channel_group);
+
+    let channels = UrlEncodedList::from(channels).into_inner();
+    let channels = if channels.is_empty() {
+        "-".to_owned()
```

Comment:
Whoops, this is not right. Empty channels are technically allowed when subscribing to channel-groups. Quoting the [REST API docs](https://www.pubnub.com/docs/pubnub-rest-api-documentation#publish-subscribe-subscribe-get-v2):

> The channel name(s) you are subscribing to. Verify that channels are comprised of valid characters. You may subscribe to mulitple channels using a comma seperator. **If subscribing to no channels (only channel groups), use a comma char (,) as a placeholder.** You may subscribe to channels, channels & channel groups, or just channel groups.

Emphasis added. The "placeholder" value can be considered a two-element list with both elements being empty strings.

Reply:
I think I meant to use a comma there (`,`) instead if the dash. Would that make more sense?

Reply:
Yes, that will fix it! Thanks.